lib/paper_trail/version_concern.rb

```diff
@@ -417,7 +417,14 @@ def reify_has_many_through(associations, model, options = {})
       associations.each do |assoc|
         next unless assoc.klass.paper_trail_enabled_for_model?
         through_collection = model.send(assoc.options[:through])
-        collection_keys = through_collection.map { |through_model| through_model.send(assoc.foreign_key) }
+
+        # if the association is a has_many association again, then call reify_has_manys for each through_collection
+        if !assoc.source_reflection.belongs_to? && through_collection.present?
+          through_collection.each { |through_model| reify_has_manys(through_model,options) }
+          return
+        end
+
+        collection_keys = through_collection.map { |through_model| through_model.send(assoc.association_foreign_key)}
```

Comment:
In the example of Document, Sections and paragraphs:
The through_model is a object of Section.

What we are trying to do is get the list of all the paragraphs associated to the section by calling:
section.paragraph_ids

If we do Section.reflect_on_association(:paragraphs).foreign_key, it returns 'section_id' and if we do Section.reflect_on_association(:paragraphs).association_foreign_key, it returns 'paragraph_id'. So we should be using 'association_foreign_key' instead of 'foreign_key'.

This was not an issue so far because in a 'belongs_to' association assoc.association_foreign_key and assoc.foreign_key are the same.